### DISCUSSION TITLE
How to use two formulas with conflicting binary names

### DISCUSSION THREAD

**Discussion by crunchtime-ali** [Homebrew/discussions#98] [Everyday usage] [answered]
I recently added the [bit-git formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/bit-git.rb) to homebrew-core which I symlink to `bit` because there was already a [formula for bit](https://github.com/Homebrew/homebrew-core/blob/master/Formula/bit.rb).

Some `bit-git` users which also use the other `bit` formula have been asking how they can install both formula (the bit-git one has a `conflicts_with` key).
Is that possible without modifying the formula or do the users have to resort to install one of them without Homebrew?

**Comment by gromgit** [comment]
There can be only one `/usr/local/bin/bit` at any one time, no matter which formula it comes from.

Users can install both formulae, but they need to `brew unlink` whichever one's already installed, before installing the other formula. `brew link/unlink` controls which formula's files are symlinked into the main `/usr/local` hierarchy, but both formulae's files can be accessed independently under `/usr/local/opt/{bit,bit-git}`.

**Reply by MultifokalHirn** [reply]
but I would say it is conceivable that brew could have a feature to work around such conflicts, through aliases or some such

**Reply by gromgit** [reply, marked as answer]
As I see it, the show-stopper is _organizational_, not technical. When you have two (or more) packages with conflicting names...
1. Who gets to keep the original name? (Everyone else has to be renamed, and when each program does something different, users may be in for a BIG surprise when `xyzzy` doesn't behave the way they thought. Any scripts/programs that expect a particular `xyzzy` may then malfunction with disastrous consequences.)
2. Who decides? (No one will be happy with this person.)
3. What happens when a new package introduces an unforeseen conflict?
4. What happens when that new package is in a third-party repo?

"Closed-development" Linux distros like Debian solve [1] through [3] by being "masters of their domains". They decide internally which packages go into the distros, so the name of every binary in an "all-packages" install is known before release, and they can decide which programs get renamed _and make those decisions stick_, but as far as I know, none of them have a solution for [4].

"Open-contribution" Linux distros like Arch face exactly the same problem as Homebrew. As far as I know, Arch chose to punt the problem to the user ("_someone's_ gotta rename _something_, we won't let you install both"). Homebrew chose (in my mind) a more useful path: you can install conflicting formulae (say, `foobar` and `burtle`), but you'll have to `brew unlink` them first, then you can decide that, say, the `xyzzy` program from `foobar` is more useful to you, so `brew link foobar` to make that `xyzzy` available without full pathname, and refer to the other one as `${HOMEBREW_PREFIX}/opt/burtle/bin/xyzzy`, or temporarily prepend `${HOMEBREW_PREFIX}/opt/burtle/bin` to your `PATH`, or `brew unlink foobar; brew link burtle` as and when your needs change.

If you can come up with an alternative solution that sidesteps [1] through [3], without negating Homebrew's "open-contribution" commitment, please file a PR to improve the situation.

**Reply by ageisen2000** [reply]
one example im running into, I'm developing on Azure AKS and sidero omni currently. Both install conflicting versions of kubelogin...

**Reply by gromgit** [reply]
Then, as I said, one of them will have to be used unlinked. For instance:-
```
brew install azure-aks
brew unlink azure-aks
brew install sidero-omni
```
then the omni version of `kubelogin` would be in Homebrew's standard PATH, but you'd have to run Azure AKS' version of `kubelogin` as `${HOMEBREW_PREFIX}/opt/azure-aks/bin/kubelogin`.
